feat: nix support for the nix folks

Alb-O opened the pull request:
Adds Nix flake (+ supporting CI tools) to build OpenCode from source.

To enable this, a few tweaks were done to the core TS codebase that shouldn't impact non-nix builds/runtimes:

  - `packages/opencode/src/cli/cmd/tui/thread.ts`
    - Nix builds produce a standalone binary with a pre-bundled worker (`opencode-worker.js`) next to the executable. The new logic checks for this bundled worker and prefers it over the development-time `worker.ts` URL.
    - Wrap worker instantiation in try-catch to handle worker init failures. If the worker fails (e.g., missing worker bundle), falls back to an inline server instance; the binary should work even if the worker bundle is missing/incompatible with the platform.
  - `packages/opencode/src/provider/models-macro.ts`
    - Checks for `MODELS_DEV_API_JSON` env var pointing to a pre-fetched copy of the models API JSON. During Nix builds, process the pre-fetched json and inject its path via this env var. Allows builds to complete without network calls.
  - `packages/sdk/js/src/gen/client/utils.gen.ts`
    - Adds string coercion for `config.baseUrl` to handles cases where `baseUrl` might be a URL object or other non-string type during Nix builds, where bundling/compilation may preserve different object types.
  - `packages/opencode/src/tool/bash.ts`
    - Previously used relative URLs in compiled bundles causing the runtime not being able to resolve wasm assets if they were packaged differently. We now resolve those URLs to absolute paths regardless of packaging method.

GitHub workflow CI:

- `update-nix-hashes.yml` triggers whenever updates to package/lock files are pushed/merged.
- The job forces a rebuild of the node_modules FOD to get the new hash, and commits the result to `hashes.json`

---

relates to:
#3540
#3739
#462
#1903

Alb-O committed:
nix flake: initial flake and patches for ts

Alb-O committed:
nix flake: bundle TUI worker for nix builds

lrittel commented:
I'm in no position to properly review this, but I'd have two suggestions:

- the version is hardcoded in the `flake.nix`, maybe it's an idea to `builtins.fromJSON` the version from a `package.json` file (e.g. `packages/opencode/package.json`)?
- maybe a GitHub Actions workflow so that breakages can be detected early? These tend to eat a few minutes, though.

Alb-O committed:
nix flake: get the version from package.json instead of hardcoded

Alb-O committed:
nix flake: add github workflow for automatically updating hashes

Alb-O committed:
nix flake: support arm64 builds

rekram1-node commented:
thanks for ur feedback @lrittel 

I think some people in the discord are also taking looks at it.

Alb-O committed:
nix flake: workflow cleanup and tweaks

Alb-O commented:
Currently working on the github workflow to automatically update hashes of build deps in flake.nix. Any preferences around this, i.e. triggers or commits from bot?

Alb-O committed:
nix flake: canonicalize node modules for reproducibility

nix: improve flake tooling and workflow robustness

- Add fetch-depth: 0 to workflow checkout to enable git rebase
- Run canonicalizer directly from source instead of copying
- Support dynamic branch names in update workflow using GITHUB_REF_NAME

The workflow previously failed on rebase because shallow clones lack
history. Running the canonicalizer in-place keeps the build tree cleaner
and avoids drift. Dynamic branch support allows the workflow to run on
any branch instead of hardcoding 'dev'.

nix: add commented-out workspace symlink rewriting option

Current .bun/node_modules rewriting provides sufficient stability.
Added optional extension to rewrite all workspace-level symlinks
through the canonical tree if needed in the future.

rekram1-node commented:
prolly the same way we do it for automatic version updates of package.json files? woudl that work

Alb-O commented:
Oh boy it was fun trying to get the node modules perfectly reproducible with Bun’s symlink tree, think I found a way though, it builds in github CI and on my local machine now. Next up is support for darwin builds and configuring the workflow automation

rekram1-node commented:
amazing thank you @Alb-O

Alb-O committed:
nix flake: refactor/split bun-build script into script/nix dir

Alb-O committed:
nix flake: refactor/split bun-build script into script/nix dir

Alb-O committed:
nix flake: completed modularization refactor

Alb-O committed:
nix flake: fix/support aarch64

Alb-O committed:
nix flake: refactor hash collection to unified json file

Alb-O committed:
nix flake: ensure optional packages list is up-to-date

Alb-O committed:
nix flake: rewrite of workflow to hopefully be more robust

Alb-O committed:
nix flake: handle version changes of package deps

Alb-O committed:
nix flake: extra fallback to allow script to get actual hash using `nix-store --realise`

Alb-O committed:
nix flake: trigger workflow on package/bun.lock changes

nix flake: cleanup

delafthi commented:
Thank you for your work on this! It looks amazing, and I'm really looking forward to it. Resources like this are incredibly helpful as I work on improving the nixpkgs release - something I've been struggling with lately.

Alb-O committed:
nix flake: initial darwin build support

Alb-O committed:
nix flake: workflow: use 4 native system runners

Alb-O committed:
nix flake: pin node_modules metadata so derivation no longer drifts between CI and local

gabrielrinaldi commented:
Thank you for the work on this, can't wait to have a proper non-hacked/patched solution for nix!

Alb-O commented:
@rekram1-node Ready for reviewing/testing now (especially Mac+Nix users would be good)

`nix run github:Alb-O/opencode/nix-support`

For specific architectures:

`nix run github:Alb-O/opencode/nix-support#packages.aarch64-linux.default`
`nix run github:Alb-O/opencode/nix-support#packages.x86_64-linux.default`
`nix run github:Alb-O/opencode/nix-support#packages.aarch64-darwin.default`
`nix run github:Alb-O/opencode/nix-support#packages.x86_64-darwin.default`

delafthi commented:
> @rekram1-node Ready for reviewing/testing now (especially Mac+Nix users would be good)

On `aarch64-darwin`

```sh
 > nix run github:Alb-O/opencode/nix-support
error: hash mismatch in fixed-output derivation '/nix/store/09wqyqc7dvm6s8va3hbwc176py3wrnbi-opencode-node_modules-1.0.39.drv':
         specified: sha256-AwTmHj8CvTBllQUnP4atM58b0bVMjvXi9yRZt1YTVbE=
            got:    sha256-Vu1drIvA5tImD4ajk152AE4HQ6yGE/8SaEgsKwgC4zM=
error: Cannot build '/nix/store/pl7kjk0sb72vq2qhzvwyvxp2fm5w2px6-opencode-1.0.39.drv'.
       Reason: 1 dependency failed.
       Output paths:
         /nix/store/r8sqgkwycljr82kfnpjmqgww9yvpamzl-opencode-1.0.39
```

It seems something's not right. Running it with 

```
 > nix run github:Alb-O/opencode/nix-support#packages.aarch64-darwin.default
error: hash mismatch in fixed-output derivation '/nix/store/09wqyqc7dvm6s8va3hbwc176py3wrnbi-opencode-node_modules-1.0.39.drv':
         specified: sha256-AwTmHj8CvTBllQUnP4atM58b0bVMjvXi9yRZt1YTVbE=
            got:    sha256-6tHiBvO1ke1m7iEhU2WBqDlxleWdBKtZjhHz/Qh6Fmc=
error: Cannot build '/nix/store/pl7kjk0sb72vq2qhzvwyvxp2fm5w2px6-opencode-1.0.39.drv'.
       Reason: 1 dependency failed.
       Output paths:
         /nix/store/r8sqgkwycljr82kfnpjmqgww9yvpamzl-opencode-1.0.39
```

yields a different hash.

`x86_64-darwin` on the other hand worked 🎉

Alb-O commented:
> yields a different hash.

Thank you for testing, these tiny, tiny, differences between the GitHub runner environment and actual user hardware that results in a different hash is driving me a bit mental, but I'll look into it nonetheless. I had it working well on linux x86_64 (exact same hashes between runner and my machine), so maybe an OS/darwin issue? Would be good to know if linux builds are working consistently

Edit: I just realized you meant there was a hash difference between the two different commands which really should give the exact same result. Hm this is a good clue

Alb-O committed:
nix flake: pin platform specific optional packages

Alb-O committed:
nix flake: only npm pack when target bundle is absent

Alb-O commented:
Ok, may have found the culprit, hopefully this extra optional package pinning solves this. Please test again, same build/run commands as before

Alb-O committed:
nix flake: add hash guard workflow  and add better metadata to auto commits

Alb-O committed:
nix flake: expanded triggers for the workflow, use hash metadata with trigger/update commits, gate releases via a new guard workflow that verifies hashes are current

delafthi commented:
> Please test again

The situation has reversed: `aarch64-darwin` now works, but `x86_64-darwin` fails with a hash mismatch:

```sh
> nix run github:Alb-O/opencode/nix-support#packages.x86_64-darwin.default
error: hash mismatch in fixed-output derivation '/nix/store/n8vbgmr3lsn5qixq9gk9y73i673rq3rc-opencode-node_modules-1.0.39.drv':
         specified: sha256-MTZm3ogq4XgsglNzaUCFD3RoHFsjGkaTnyoU4fEn+rI=
            got:    sha256-q1uEMWquAaV1Xd0bzKZKm+S4/uW3xpwadBQU3RPl53U=
error: Cannot build '/nix/store/0yxv492hr4zf4dyjps9m5g5m1wlvngp6-opencode-1.0.39.drv'.
       Reason: 1 dependency failed.
       Output paths:
         /nix/store/231qf56dni35fx7176vd8x5pilc9fpbi-opencode-1.0.39
```

The `x86_64-darwin` build produces the same hash on repeated runs, indicating a consistent mismatch.

ndrwstn commented:
aarch64-darwin is failing on the models.dev api.json hash as well:
```
> nix run github:Alb-O/opencode/nix-support#packages.aarch64-darwin.default
error: hash mismatch in fixed-output derivation '/nix/store/gky155qq5bc9ilwas8r43c5p41gxnhkn-api.json.drv':
         specified: sha256-Q82BpJzo/Gla3M3RHAwhATYV0NxamkdwoVL3iEUbaec=
            got:    sha256-dTUM9lSUErd28jMw/YVWwpD0mZWfTRgzfysfq5KhKjg=
```
Given that models.dev is external to this repo, I don't think you can have a hard match to api.json that will be sustainable. I think you just have to accept the non-declarativeness of fetching it and taking whatever you're given, maybe with a null fallback, since if you're trying to use opencode in an airgapped environment you probably either know what you're doing or shouldn't be doing it anyway.

delafthi commented:
I encountered this exact issue when I first packaged OpenCode for nixpkgs. The problem stems from fetching from a link which doesn't pin a specific version - whenever `api.json` gets updated on the server, you'll get a hash mismatch. In other words, the output of that `fetchContent` isn't consistent.

I solved this by packaging `models-dev` separately (see the [package definition](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/mo/models-dev/package.nix)) and then using that as a dependency.

ndrwstn commented:
I don't understand the opencode architecture choice, but it seems like (I just asked an agent; I didn't verify:) opencode will crash on first run if api.json isn't accessible (e.g. offline). As a result, it's included in the build even though it's not really needed (or even desirable, frankly).

I think the **right** answer is to fix the application to check for connectivity and cleanly error on first-run instead of forcing a download on build, but that's well beyond the scope of making a nix flake. For the moment, since it doesn't matter, you just need to pick a deterministic way of giving opencode *something* on first-run, whether you refer to nixpkgs models-dev or just pass a null.

rekram1-node commented:
@ndrwstn the macro is needed so there is a copy of available models when you run opencode

Alb-O committed:
nix flake: remove direct models.dev json fetching, replace with equivalent nixpkgs distribution

Alb-O committed:
nix flake: properly cancel previous workflow

Alb-O committed:
nix flake: test with missing hashes.json

Alb-O committed:
nix flake: verify npm packages sri

Alb-O committed:
nix flake: remove infinite loop check block (unnecessary)

Alb-O commented:
Ready to be tested again, models.dev API JSON is now sourced from nixpkgs, and I'm doing extra hash validation on other dependency packages

Same commands as before:

`nix run github:Alb-O/opencode/nix-support`

For specific architectures:

`nix run github:Alb-O/opencode/nix-support#packages.aarch64-linux.default`
`nix run github:Alb-O/opencode/nix-support#packages.x86_64-linux.default`
`nix run github:Alb-O/opencode/nix-support#packages.aarch64-darwin.default`
`nix run github:Alb-O/opencode/nix-support#packages.x86_64-darwin.default`

Alb-O committed:
nix flake: minor cleanup

Alb-O committed:
nix flake: track every asset Bun emits during compile, install them durign the installPhase

Alb-O committed:
nix flake: resolve wasm assets to absolute filepaths

nix flake: track every asset Bun emits during compile, install them durign the installPhase

delafthi commented:
Tested the current version. Success, both `x86_64-darwin` and `aarch64-darwin` worked.

Alb-O committed:
nix flake: remove (1) of the unneeded scripts

Alb-O committed:
nix flake: add `[skip ci]` to auto commit message

Alb-O committed:
Merge branch 'dev' into nix-support

Alb-O committed:
nix flake: fix commit metadata

Alb-O committed:
Merge branch 'dev' into nix-support

Commit:
nix: update hashes (branch:nix-support)

Alb-O committed:
nix flake: add more explicit sorting

Alb-O committed:
nix flake: add more explicit sorting

Commit:
nix: update hashes (branch:nix-support-dev)

Commit:
nix: update hashes (branch:nix-support-dev)

Alb-O committed:
Merge branch 'dev' into nix-support

Alb-O committed:
Merge branch 'nix-support' into nix-support-dev

Alb-O committed:
nix flake: remove a redundant builds from workflow

Alb-O commented:
Unfortunately, despite my best efforts, there are STILL sometimes random hash inconsistencies in the node_modules fixed output derivation between the github CI and local builds. The annoying part is it's VERY close to being consistent and matching up exactly with local builds... most of the time. I suspect this slight randomness comes from somewhere in bun's internal build process, and I'm not sure how to solve this. If anyone has ideas, let me know.

gabrielrinaldi commented:
This has be working for me on NixOS (apart from the rewrite bugs that I assume are unrelated), has been consistent across multiple builds (I might be getting lucky).

Alb-O commented:
> This has be working for me on NixOS (apart from the rewrite bugs that I assume are unrelated), has been consistent across multiple builds (I might be getting lucky).

This is probably because I have been manually re-running the CI on my fork immediately after the original run produced a bad linux 86x hash (I'm also using NixOS). Interestingly, if I change the workflow yml around a bit (e.g. removing some jobs/steps) then the linux hash suddenly becomes stable across runs... but the inconsistency problem shifts over to the MacOS hashes instead... 🤪

Alb-O committed:
Merge branch 'dev' into nix-support-dev

Alb-O committed:
nix flake: remove commit metadata info

patrickjcasey commented:
I just tested `nix run github:Alb-O/opencode/nix-support` on x86_64 Linux and it worked for me!

Commit:
Update Nix hashes

Commit:
Update Nix hashes

Alb-O commented:
Here's proof of the CI run inconsistencies:

[auto-commit on nix-support-dev branch](https://github.com/Alb-O/opencode/commit/22b928fb2085c64efc0f9cc443ade812d9f2ca67)

[auto-commit on nix-support branch](https://github.com/Alb-O/opencode/commit/bbfa63da7b44ed5bc8f6d51d44139b7e13a7451c)

The parent commit for both of these are identical (3c160aa), exact same git history up until this point. Yet something is still touching the node_modules result hashes in a non-deterministic way. In this specific case, the linux hashes changed while the mac hashes remained stable...

Ninja3047 commented:
maybe you can add something to the workflow that saves the node_modules from the store as an artifact so they can be compared locally using something like diffoscope. You can do something like tar.gz up the folder and upload it via actions/upload-artifact with a short expiry window

FWIW this is what i'm getting
```
error: hash mismatch in fixed-output derivation '/nix/store/ndsb5q1mhcss6x7qwz52jwrpgn0130xx-opencode-node_modules-1.0.58.drv':
           likely URL: (unknown)
            specified: sha256-+dsSECijB/O6dA87Rdm8mcSrPn7FoMaNM4pzgouyFn0=
                  got: sha256-U2bUPxnKrUN6pRIDBIiyrg+CqghaoneSqY4BI5zXs5w=
        expected path: /nix/store/hw9psi0abc8j4pd8n8kl2jbxbph5c5id-opencode-node_modules-1.0.58
             got path: /nix/store/h6p7lqckj4v2d8qhhlima5qkph3wdls4-opencode-node_modules-1.0.58
error: 1 dependencies of derivation '/nix/store/qrlixkk1pg2k9xkjpdsyxk2pn9y153lg-opencode-1.0.58.drv' failed to build
error: 1 dependencies of derivation '/nix/store/nlbh1jymp215cmlrxm1913hwi3wyb1g8-nix-shell-env.drv' failed to build
```
but the `nix-support-dev` branch works for me

Alb-O committed:
nix flake: use '--linker=isolated' in bun install

Alb-O committed:
nix flake: revert bundle-utils

Alb-O committed:
nix flake: revert verify sha script

Alb-O committed:
nix flake: revert excessive optional metadata verification

Alb-O committed:
nix flake: attempt full regen of hash file

Alb-O committed:
nix flake: remove all optional metadata/packages handling

Commit:
Update Nix hashes

Alb-O committed:
Merge branch 'dev' into nix-support-dev

Commit:
Update Nix hashes

Alb-O committed:
nix flake: test without isolation flags on opentui/watcher

Commit:
Update Nix hashes

Alb-O committed:
Merge branch 'nix-support-dev' into nix-support

Commit:
Update Nix hashes

Alb-O committed:
Add workflow to snapshot node modules

Alb-O committed:
Trigger snapshot workflow on debug branches

Alb-O committed:
Fix artifact glob for snapshot workflow

Alb-O committed:
Trigger snapshot run #2

Alb-O committed:
Normalize bun .bin directories for determinism

Alb-O committed:
Trigger snapshot run #3

Alb-O committed:
Publish hash metadata separately

Alb-O committed:
Update Nix node_modules hashes

Alb-O committed:
Merge branch 'dev' into nix-support-dev

Alb-O committed:
Merge branch 'debug/node-modules-hash' into nix-support-dev

Commit:
Update Nix hashes

Alb-O committed:
Merge branch 'dev' into nix-support-dev

Alb-O committed:
nix flake: remove debug and cleanup

Alb-O commented:
Hash inconsistency issues seem to be finally resolved, pinpointing the issue to symlinked binaries in `node_modules/.bun` and `.bin`, and normalizing these paths during nix builds. Looking nice and stable now 🤞

Alb-O committed:
Merge branch 'dev' into nix-support

Alb-O committed:
Merge branch 'dev' into nix-support

delafthi commented on nix/node-modules.nix:
```diff
@@ -0,0 +1,50 @@
+
+  buildPhase = ''
+    runHook preBuild
+    export HOME=$(mktemp -d)
+    export BUN_INSTALL_CACHE_DIR=$(mktemp -d)
+    bun install \
+      --frozen-lockfile \
+      --ignore-scripts \
+      --no-progress \
+      --linker=isolated
```
With `--cpu="*"` and `--os="*"` I'm able to prevent platform specific hashes.

Alb-O replied:
wow. assuming no issues for mac people, this is a godsend

delafthi commented on nix/opencode.nix:
```diff
@@ -0,0 +1,109 @@
+    runHook postInstall
+  '';
+
+  postFixup = lib.optionalString stdenvNoCC.hostPlatform.isLinux ''
```
According to the build script for the AUR `fzf` and `ripgrep` are runtime deps. See https://github.com/sst/opencode/blob/f6b3ffaf64f2cb361f940b50029771f8ef12947e/packages/opencode/script/publish.ts#L89

So, I added them to path in `nixpkgs`

```nix
  postInstall = ''
    wrapProgram $out/bin/opencode \
      --prefix PATH : ${
        lib.makeBinPath [
          fzf
          ripgrep
        ]
      }
  '';
```

delafthi commented on nix/opencode.nix:
```diff
@@ -0,0 +1,109 @@
+
+  postFixup = lib.optionalString stdenvNoCC.hostPlatform.isLinux ''
+    wrapProgram $out/bin/opencode \
+      --set LD_LIBRARY_PATH "${lib.makeLibraryPath [ stdenv.cc.cc.lib ]}"
```
Question here: Is this necessary, or is that still a remnant from previous derivations?

Alb-O replied:
`patchelf --print-needed result/bin/opencode` prints just `libc.so.6`, so doesn't look necessary, you're right

Alb-O committed:
nix flake: add fzf/ripgrep runtime deps to the wrapper

Alb-O committed:
nix: fetch bun deps for all platforms

Alb-O committed:
nix flake: drop LD_LIBRARY_PATH

Alb-O committed:
Merge branch 'dev' into nix-support

Commit:
Update Nix hashes

Alb-O committed:
Merge branch 'dev' into nix-support

Commit:
Update Nix hashes

Alb-O committed:
nix flake: update readme to include nix method

rekram1-node commented:
@Alb-O is the goat

rekram1-node commented:
/oc run `./packages/sdk/js/script/build.ts` and then commit the changes

opencode-agent[bot] committed:
build: update SDK client utils after build

Mic92 commented:
Triggeres the same bug on macOS that I have seen before when doing my own packaging. I opened https://github.com/sst/opencode/issues/4575 for discussion.

ck3mp3r commented:
Something is not quite right... this is what happens on an M3 MBP, how do you generate the hash for the fixed output derivation? Only for one architecture?

```
❯ nix run github:sst/opencode
[1/0/2 built, 5 copied (246.6 MiB), 91.0 MiB DL] building opencode-node_modules-1.0.108 (inst
error: hash mismatch in fixed-output derivation '/nix/store/xmy6279nf207pidmm3cxnq24zhr89l0k-opencode-node_modules-1.0.108.drv':
         specified: sha256-m7hL9Uzqk+oa2/FtgkzEPgi+m/VZP1SvjpgYHNjiS1c=
            got:    sha256-83nrKWOCUJhSHqtSApFuXxggEwO86rIWhy5d2vtfz4Y=
error: Cannot build '/nix/store/jjc87gkbwkhdajzca6g653ay0w20kppm-opencode-1.0.108.drv'.
       Reason: 1 dependency failed.
       Output paths:
         /nix/store/lxacjg2y1gcyqbbqb7cbhpz6axl7kq14-opencode-1.0.108

```

gabrielrinaldi commented:
I think you just got in between the job that generates the hash for nix not having run yet, at least for me if I wait a little it usually works, the release is out before the new hash is updated. (this is my understanding of the process by looking at the commits and personal experience)

rekram1-node commented:
I don't know anything about nix but I know why it keeps behaving weird to some degree Ill try to fix it today, think it just involves moving stuff around

ck3mp3r commented:
I guess for now I will stick to my own nix build of opencode until this one is more stable.

```
❯ nix run github:sst/opencode
[1/0/1 built] building opencode-1.0.110 (fixupPhase): _: line 2: patchelf: command not found

```

Note: it did build after quite a while even with the error. My approach is to use the pre-built binaries from the release page as non flake inputs, which avoids the entire build and also avoids any fixed output derivation hashes.
I also symlink the contents instead of copying them on first run... feel free to have a look if you want `github:ck3mp3r/flakes?dir=opencode`. I also change the cache directory to something else like `~/.cache-nix/opencode` I run opencode via a wrapper script with the same name.
I'm sure my version needs some optimisation, but it does the trick for me reproducibly and reliably. Also, because my workflow caches the entire output as derivations on my nix cache, the machines consuming this never need to run any builds, nor do they need to interact with npmjs et al.

Alb-O commented:
Yes, you can simply fetch/patch release tarballs. But this flake is targeted mainly for contributor usage where building and running the dev work tree which doesn't have a release yet is a must. It also helps out the nixpkgs maintainers who prefer packages to be built from source (and cached)

ck3mp3r commented:
> Yes, you can simply fetch/patch release tarballs. But this flake is targeted mainly for contributor usage where building and running the dev work tree which doesn't have a release yet is a must. It also helps out the nixpkgs maintainers who prefer packages to be built from source (and cached)

yes, this flake should definitely become the source of the upstream nix package.
